### DISCUSSION TITLE
Failing Test: invariant_protocolMustHaveMoreValueThanTotalSupply

### DISCUSSION THREAD

**Discussion by pherm0007** [Cyfrin/foundry-full-course-cu#2934] [Q&A]
'm encountering a persistent failure in the invariant test invariant_protocolMustHaveMoreValueThanTotalSupply while testing the DSCEngine contract using Foundry. The test fails due to an arithmetic underflow or overflow in the burnDsc(uint256) function.

<img width="1337" alt="Screenshot 2024-11-27 at 09 45 16" src="https://github.com/user-attachments/assets/d3b62c45-86da-4404-ac25-d464e718fd6c">
<img width="1680" alt="Screenshot 2024-11-27 at 09 45 54" src="https://github.com/user-attachments/assets/41ebc365-3dec-4f01-b5f0-25bdd66daf36">

**Comment by EngrPips** [comment]
Hello @pherm0007, Your stack trace shows that when the Fuzzer called the `burnDSC` function on your `DSCEngine`, that resulted in an `arithmetic underflow or overflow` error. Please look into that.

**Reply by pherm0007** [reply]
i. am getting this error 
<img width="1680" alt="Screenshot 2024-11-27 at 10 28 56" src="https://github.com/user-attachments/assets/6adc33d7-fe3e-4f89-bc2b-ed9ff90804ef">

**Comment by pherm0007** [comment]
here is an error i am getting now 
<img width="1680" alt="Screenshot 2024-11-27 at 10 28 56" src="https://github.com/user-attachments/assets/d05c0448-ba49-4177-84f2-2137d63e851a">

**Reply by EngrPips** [reply]
The same thing is happening for `redeemCollateralForDSC` Please look into it.

**Reply by heheboii11** [reply]
Hey @EngrPips, I guess there are some error with the code that patrics used!!!
```Here is what i believe that is wrong!. Please let me know if I am wrong```
 In the patrics code as shown in the lesson, He used ```_getCollateralFromSeed``` to get a ```collateral address```, and he uses that address to deposit the collateral and to redeem it. But the main problem here is that, when doing fuzz tests, it calls random functions with random inputs which causes it might choose ```weth``` to deposit collateral and call redeem collateral with ```wbtc```  as collateral to redeem. This causes to break the test right ??

I got stuck in the same error as shown above and feel so exhausted trying to solve this. Please help me to clarify the above doubt.

**Reply by EngrPips** [reply]
I will advise that you check the Patrick code here https://github.com/Cyfrin/foundry-defi-stablecoin-cu/blob/main/test/fuzz/failOnRevert/StopOnRevertHandler.t.sol  to see how he handled such situation that you stated above.

**Reply by heheboii11** [reply]
I have gone through the code, I understood it will not be a problem, because the collateral amount will return as zero and redeeem will just return, which is fine

**Comment by heheboii11** [comment]
Coming to @pherm0007 , please trying changing the code little bit because even if you see in the updated code, as @EngrPips provided above, where in the code the ```mint handler``` got commented out. Means this wont get executed. This is because, if you first deposit the collateral, mint some DSC and then if you try to redeem the collateral, in this case even though you are setting the maxredeem to be the amount of collateral available with user but you have already minted some DSC which makes the initial health factor to reduce after minting. And now you are trying to redeem collateral, where maxredeem is the amount of the collateral you have deposited so if you try to redeem all the collateral, the health factor breaks. so the max redeem set by patrics feels wrong and needs to be modified. here is how i have done 
```solidity
function Collateralredeem(uint256 CollateralSeed, uint256 amountofCollateral) public {
        ERC20Mock collateral = getCollateralSeed(CollateralSeed);
        uint256 Dscminted = engine.getDScmintedbyUser(msg.sender);
        uint256 DScValueInCollateral = engine.getUSDValueintoken(address(collateral), Dscminted);
        uint256 Collateralvalue = engine.getCollateralwithUser(address(collateral), msg.sender);
        if (Collateralvalue == 0) {
            return;
        }
        uint256 MaxRedeem = Collateralvalue - (DScValueInCollateral * 2);
        console.log("maxredeem:", MaxRedeem);
        /// in here we have tweek a bit that we have to check DSC that are already minted and deduct from maxAmount

        amountofCollateral = bound(amountofCollateral, 0, MaxRedeem);
        if (amountofCollateral == 0) {
            return;
        }
        vm.startPrank(msg.sender);
        ERC20Mock(address(coin)).approve(address(engine), Dscminted);
        engine.redeemcollateral(address(collateral), amountofCollateral);
        vm.stopPrank();
    }
```
I have subtracted that amount because the DSCengine will always be 200% overcollaterised
 @EngrPips can you please confirm if the above id vaild !!

**Reply by EngrPips** [reply]
I think this makes sense, Have you tested it to see if it works.

**Reply by heheboii11** [reply]
I haven't but will try it later today as there are some other bugs in my code rn!! Will update here later if it works 
But @pherm0007, can you please makes the changes mentioned above and try the test if it works and updates us with the results?

**Reply by EngrPips** [reply]
Alright.

**Reply by pherm0007** [reply]
@heheboii11  thanks for helping me out man..... @EngrPips  thanks for always coming thru

**Reply by EngrPips** [reply]
Anytime, Friend. Keep crushing it.

**Reply by heheboii11** [reply]
Hey @pherm0007, did the solution that i have mentioned above worked ?? Just wanted to know, if it is not correct please provide the changes you made!!!
